Masking and moments tutorial

e-koch opened the pull request:
Added tutorial covering:
* Noise calculation and identifying signal
* Creating and applying signal masks
* Calculating moment maps from the masked cube

Here's the binder link (currently pointing to my fork):
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/e-koch/tutorials/masking_and_moments?filepath=masking_and_moments%2Fmasking_and_moments_guide.ipynb)

e-koch commented:
The region rejection code with dask arrays works fine for me locally but seems to hang on binder. Unsure of why yet.

e-koch commented:
* [ ] Explain what the `include` and `compute` statements are used for
* [ ] Why +1 in indexing.
* [ ] Redundant ax plot WCS vs pixels (clean-up)

keflavich commented on masking_and_moments/masking_and_moments_guide.ipynb:
```diff
@@ -0,0 +1,918 @@
+    "\n",
+    "Another common convention is to define the full-width-half-max line width:\n",
+    "$$\n",
+    "\\textrm{FWHM}=\\sqrt{8 {\\rm ln} 8} \\sigma_v \\ \\approx 2.35 \\sigma_v\n",
```
```suggestion
    "\\textrm{FWHM}=\\sqrt{8 {\\rm ln} 2} \\sigma_v \\ \\approx 2.35 \\sigma_v\n",
```

e-koch replied:
Had it fixed locally but hadn't pushed it at the time.

keflavich reviewed (approved):
This is awesome, let's merge yesterday.

e-koch commented:
* [ ] Add in scipy equivalent masking steps (seems much faster on small cubes)

keflavich commented:
This raises a bigger question: Should we be encouraging people to  - and enabling them to - use scipy tools rather than dask tools for cubes that fit in memory?  I'm inclined to say yes, but that could be a lot of work - let's discuss when we can

e-koch commented:
It's not that much "duplicated" code, so I was going to keep it just commented out. I'll add it in today and we can discuss how to present one or both tomorrow.

keflavich commented:
I meant more generally, when we have methods implemented on cubes: we should, ideally, make a "smart" choice between dask vs scipy for each cube

e-koch commented:
@keflavich -- This is good to merge from my end. But let me know if anyone trying the tutorial out has more comments to add beforehand.

abulatek commented:
Hi all! The tutorial looks great! I really appreciated the added discussion and the addition of the `scipy.ndimage` steps (it helped me troubleshoot my implementation of `ndimage`). I have one quick question about the `num_pixels` calculations in the **Mask labels and indexing** section. In the `scipy.ndimage` implementation, for example, the code reads:

```
num_pixels_in_high_snr_mask = nd.sum(high_snr_mask,
                                     labels=low_snr_mask_labels,
                                     index=range(1, num_labels + 1)) # +1 offset for mask labels

# Repeat for the high signal mask.
num_pixels_in_low_snr_mask = nd.sum(low_snr_mask,
                                    labels=low_snr_mask_labels,
                                    index=range(1, num_labels + 1)) # +1 offset for mask labels
```
The `labels` keyword is the same in both pixel summing steps, but should the keyword be `high_snr_mask_labels` in `num_pixels_in_high_snr_mask` (we don't actually retrieve `high_snr_mask_labels`—should we be doing this?)? Also, should the comment before the second sum read "Repeat for the low signal mask"? Thanks very much, and I apologize if I'm misunderstanding how the `labels` keyword is getting used!

e-koch commented:
Thanks, @abulatek! I've added some text to clarify. Let me know what you think.

We do want to use `low_snr_mask_labels` for both because we want to use the same mask regions to calculate the pixel sums in each. All we need to know is how many `high_snr_mask` pixels fall within each label in `low_snr_mask_labels`, with the idea that regions with few high SNR pixels are more likely to be spurious noise features.

The "Repeat for the low signal mask" was definitely a typo. I've fixed that.

abulatek commented:
Awesome! I think the text you added makes it a lot clearer. Thanks for the explanation!
